### ISSUE TITLE
Add explicit HTTP2 over stdio backend (similar to rclone)

### ISSUE THREAD

**Issue by julianbrost** [restic/restic#2770]
Output of `restic version`
--------------------------

restic 0.9.6 compiled with go1.14 on linux/amd64

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

With the rclone backend, restic already implements accessing a repository by communicating with another process using HTTP2 over stdin/stdout. While implemented for accessing all kinds of remote storage using rclone, this can also be used in different ways, for example to connect to a SSH server which then serves a repository using a force command.

Currently, this can be done with a restic command line like for example:

    restic -o rclone.program='ssh restic@remote' -o rclone.args='' -r rclone: init

This works totally fine, however, it feels quite hacky to actually do this. Therefore I propose adding a similar second backend, that would allow the same with a nicer command line like for example:

    restic -r stdio:'ssh restic@remote' init

I don't think this should add too much code (one could be implemented as a special-case of the other), but it would greatly improve the user experience in these use cases.

What are you trying to do?
--------------------------

See for example https://ruderich.org/simon/notes/append-only-backups-with-restic-and-rclone, but in a way that feels less hacky.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

Not today, but it already saved my ass a few times ;)

**Comment by greatroar** [comment]
I don't find "it feels hacky" a compelling reason to introduce alternative syntax for existing functionality. The use case is to still run rclone, except on a different host via SSH, so the current name is apt and everything works fine. The syntax of such an advanced setup being somewhat complex is to be expected.

**Comment by julianbrost** [comment]
Sure, the current method works and I'm not aware of something else than rclone implementing the protocol right now.

The currently rclone backend is clearly built with running rclone locally in mind. I can think of some options that would make limited to no sense for this but could be useful with a more generic backend in mind. For example an option to specify that the program should be restarted once it exits to allow to recover from a failed connection (see also #2381 which kind of was the inspiration for this feature request) or one to specify that multiple instances of the process should be started in parallel, since SSH is single-threaded and can easily become the bottleneck in such a setup.

If this gets blessing from the maintainers, I'd also be willing to contribute code for this, so that this won't become one of these feature requests lying around that would be nice to have, but nobody really cares enough to actually implement.

**Comment by streaps** [comment]
> I don't find "it feels hacky" a compelling reason to introduce alternative syntax for existing functionality.

Improving usability and making tasks easier for the user is not a good reason?

**Comment by MichaelEischer** [comment]
I don't really see a benefit from having an additional backend for a few special cases. In particular, that would likely mean duplicating a significant part of the code of the rclone backend. Special behavior such as reconnecting would be opt-in independent of whether it uses the `rclone` or `stdio` backend. Thus, I'd only expect only duplicated code, but no real benefit.
